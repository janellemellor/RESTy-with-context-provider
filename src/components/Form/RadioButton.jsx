import React from 'react';
import PropTypes from 'prop-types';

export const RadioButtonGroup = ({ children, name, onInputChange }) => {
  const radioButtons = React.Children.map(children, child => {
    return React.cloneElement(child, {
      name,
      onInputChange 
    });
  });

  return (
    <>
      {radioButtons}
    </>
  );
};

export const RadioButton = ({ name, value, onInputChange }) => (
  <>
    <input type="radio" name={name} value={value} onInputChange={onInputChange} />
    <label htmlFor={value}>{value} </label>
  </>
);


RadioButtonGroup.propTypes = {
  children: PropTypes.node,
  name: PropTypes.string.isRequired,
  onInputChange: PropTypes.func.isRequired,
};

RadioButton.propTypes = {
  name: PropTypes.string.isRequired,
  onInputChange: PropTypes.func.isRequired,
  value: PropTypes.string
};

