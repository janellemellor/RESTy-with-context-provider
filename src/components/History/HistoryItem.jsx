import React from 'react';
import PropTypes from 'prop-types';

const HistoryItem = ({ url, method }) => (
  <section>
    <p>{method}</p>
    <p>{url}</p>
  </section>
); 

HistoryItem.propTypes = {
  url: PropTypes.string.isRequired,
  method: PropTypes.string.isRequired
};

export default HistoryItem;
