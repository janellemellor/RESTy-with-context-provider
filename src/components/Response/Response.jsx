import React from 'react';

const Response = ({ response }) => (
  <pre>
    {JSON.stringify(response, null, 2)}  
  </pre>
);

export default Response;
