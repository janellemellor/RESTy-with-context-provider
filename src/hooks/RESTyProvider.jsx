import React, { createContext, useState, useContext } from 'react';
import { fetchRequest } from '../services/fetchRequest.js';

const RESTyContext = createContext();

// eslint-disable-next-line react/prop-types
export const RESTyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [jsonBody, setJsonBody] = useState('');
  const [response, setResponse] = useState({});

  const onChange = ({ target }) => {
    const setInputsFactory = {
      url: setUrl,
      method: setMethod,
      jsonBody: setJsonBody
    };

    setInputsFactory[target.name](target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();
    fetchRequest(url, method, jsonBody)
      .then(res => {
        setResponse(res);
      }); 
  };

  const context = {
    url, 
    method, 
    jsonBody, 
    response, 
    onChange, 
    onSubmit
  };
  
  return (
    <RESTyContext.Provider value={context} >
      {children}
    </RESTyContext.Provider>
  );
};

export const useRESTy = () => {
  const context = useContext(RESTyContext);
  return context;
};




