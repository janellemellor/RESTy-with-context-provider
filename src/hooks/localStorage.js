import { useState, useEffect } from 'react';

const fetchState = (key, initialValue) => localStorage.getItem(key) ? JSON.parse(localStorage.getItem(key)) : initialValue;

export const useLocalStorage = (key, initialValue) => {
  const [state, setState] = useState(fetchState(key, initialValue));

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(state));
  }, [state]);

  return [state, setState];
};
