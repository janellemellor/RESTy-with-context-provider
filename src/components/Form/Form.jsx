import React from 'react';
import PropTypes from 'prop-types';
import { RadioButtonGroup, RadioButton } from './RadioButton';

const Form = ({ onSubmit, url, onUrlChange, onInputChange, jsonBody, onJsonBodyChange }) => (
  <form onSubmit={onSubmit}>
    <fieldset>
      <input name="url-input" type="text" value={url} onUrlChange={onUrlChange} />
      <RadioButtonGroup name="methods" onInputChange={onInputChange} >
        <RadioButton value="GET" />
        <RadioButton value="POST" />
        <RadioButton value="PUT" />
        <RadioButton value="PATCH" />
        <RadioButton value="DELETE" />
      </RadioButtonGroup>
      <button>Go!</button>
      <textarea name="jsonBody" value={jsonBody} onJsonBodyChange={onJsonBodyChange}> </textarea>
    </fieldset>    
  </form>
);

Form.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
  onUrlChange: PropTypes.func.isRequired,
  onInputChange: PropTypes.func.isRequired, 
  jsonBody: PropTypes.string,
  onJsonBodyChange: PropTypes.func
};

export default Form;
