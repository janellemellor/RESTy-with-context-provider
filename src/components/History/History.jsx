import React from 'react';
import HistoryItem from './HistoryItem';
import { useRESTy } from '../../hooks/RESTyProvider';

const History = () => {
  const { history } = useRESTy();
  
  const requestHistory = history.map(
    (item, i) => (
      <li key={i}>
        <HistoryItem {...item} />   
      </li>
    ));
  
  return (
    <ul>
      {requestHistory}
    </ul>    
  );
};

export default History;
