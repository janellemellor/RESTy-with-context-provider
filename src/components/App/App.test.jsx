import React from 'react';
import { shallow } from 'enzyme';
import { RESTyProvider } from '../../hooks/RESTyProvider';
import Form from '../Form/Form';
import Response from '../Response/Response';
import History from '../History/History';

describe('App component', () => {
  it('renders App', () => {
    const wrapper = shallow(
      <RESTyProvider >
        <Form />
        <Response />
        <History />
      </RESTyProvider>);
    expect(wrapper).toMatchSnapshot();
  });
});
