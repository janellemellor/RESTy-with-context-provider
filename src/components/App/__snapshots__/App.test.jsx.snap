// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App 1`] = `
<RESTyProvider>
  <Form />
  <Response />
  <History />
</RESTyProvider>
`;
